### ISSUE TITLE
The 'User-Agent' header must be modified using the appropriate property or method.

### ISSUE THREAD

**Issue by hempe** [Azure/azure-signalr#198]
I'm using Microsoft.Azure.SignalR 1.0.0-preview1-10200, Target platform net461.

When i try to use AzureSignalR I see the following error in the logs, and it seems no connection to the hub is established.

```
[09:24:01 ERR] Failed to connect to the service.
System.ArgumentException: The 'User-Agent' header must be modified using the appropriate property or method.
Parameter name: name
   at System.Net.WebHeaderCollection.ThrowOnRestrictedHeader(String headerName)
   at System.Net.WebHeaderCollection.Set(String name, String value)
   at Microsoft.AspNetCore.Http.Connections.Client.Internal.WebSocketsTransport..ctor(HttpConnectionOptions httpConnectionOptions, ILoggerFactory loggerFactory, Func`1 accessTokenProvider)
   at Microsoft.AspNetCore.Http.Connections.Client.Internal.DefaultTransportFactory.CreateTransport(HttpTransportType availableServerTransports)
   at Microsoft.AspNetCore.Http.Connections.Client.HttpConnection.<StartTransport>d__47.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Http.Connections.Client.HttpConnection.<SelectAndStartTransport>d__44.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Http.Connections.Client.HttpConnection.<StartAsyncCore>d__41.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Http.Connections.Client.HttpConnection.<StartAsync>d__40.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.SignalR.ServiceHubDispatcher`1.<ConnectAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.SignalR.ServiceHubDispatcher`1.<ConnectAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.SignalR.ServiceConnectionBase.<StartAsyncCore>d__26.MoveNext()
```

**Comment by xscript** [comment]
@davidfowl Is this because `ClientWebSocket` in .net framework doesn't allow set `User-Agent` header [this way](https://github.com/aspnet/SignalR/blob/release/2.2/src/Microsoft.AspNetCore.Http.Connections.Client/Internal/WebSocketsTransport.cs#L48)?

**Comment by hempe** [comment]
@xscript @davidfowl 

Seems like it:
From: [WebHeaderCollection.cs](https://referencesource.microsoft.com/#system/net/System/Net/WebHeaderCollection.cs,127b05c42eb27cc6)
```
        //
        // ThrowOnRestrictedHeader - generates an error if the user,
        //  passed in a reserved string as the header name
        //
        internal void ThrowOnRestrictedHeader(string headerName)
        {
            if (m_Type == WebHeaderCollectionType.HttpWebRequest)
            {
                if (HInfo[headerName].IsRequestRestricted)
                {
                    throw new ArgumentException(SR.GetString(SR.net_headerrestrict, headerName), "name");
                }
            }
            else if (m_Type == WebHeaderCollectionType.HttpListenerResponse)
            {
                if (HInfo[headerName].IsResponseRestricted)
                {
                    throw new ArgumentException(SR.GetString(SR.net_headerrestrict, headerName), "name");
                }
            }
        }
```
And from: [_HeaderInfoTable.cs](https://referencesource.microsoft.com/#system/net/System/Net/_HeaderInfoTable.cs,bfb2597634963745)
```new HeaderInfo(HttpKnownHeaderNames.UserAgent,          true,   false,  false,  SingleParser),```

And with HeaderInfo the second parameter is requestRestricted.
From: [_HeaderInfo.cs](https://referencesource.microsoft.com/#system/net/System/Net/_HeaderInfo.cs,700a3d7373015299)
 
```
internal HeaderInfo(string name, bool requestRestricted, bool responseRestricted, bool multi, HeaderParser p) {
            HeaderName = name;
            IsRequestRestricted = requestRestricted;
            IsResponseRestricted = responseRestricted;
            Parser = p;
            AllowMultiValues = multi;
        }
```

**Comment by davidfowl** [comment]
> @davidfowl Is this because ClientWebSocket in .net framework doesn't allow set User-Agent header this way?

Yes. We don't set the user agent in ASP.NET SignalR either. It's a bit unfortunate though. We'll need to leave it out for now.

**Comment by xscript** [comment]
@davidfowl  Right. We may have to use another custom header.

@hempe It is interesting that you are trying to use Azure SignalR Service in .net framework. Would you like to share more about why you are doing this and how it goes for now?

**Comment by michaelpaulus** [comment]
@xscript I am having this issue as well, I want to use the new service but can't fully switch over to .net core because of the lack of the accessToken on the sql client for the .net standard version, I understand this will be added in 2.2, but need to use 4.7.1 until that time.

**Comment by hempe** [comment]
@xscript We are currently migrating a lot of things in our production code from net461 to standard (but not core). The application where i want to use this is a windows service loading a lot of plugins and we are trying to make a smooth (non breaking) transition, so i can't just switch to netcore. 

But... can you even have a windows service in .net core?

**Comment by xscript** [comment]
Thanks guys for sharing with us how you are using SignalR Service! This is very helpful to us. 😄 

@hempe Have you checked out [hosting ASP.NET Core in a Windows Service](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/windows-service?view=aspnetcore-2.1)?

**Comment by hempe** [comment]
@xscript Haven't seen that. Thanks for the info. Might be useful in the future.

**Comment by xscript** [comment]
Fixed with the latest Service SDK (1.0.0).
